metric/emitter/prometheus.go

```diff
@@ -0,0 +1,188 @@
+	buildsStarted     prometheus.Counter
+	buildsFinished    prometheus.Counter
+	buildsSucceeded   prometheus.Counter
+	buildsErrored     prometheus.Counter
+	buildsFailed      prometheus.Counter
+	buildsAborted     prometheus.Counter
+	buildsFinishedVec *prometheus.CounterVec
+	buildDurationsVec *prometheus.HistogramVec
+}
+
+type PrometheusConfig struct {
+	Address string `long:"prometheus-address" description:"Address to listen on to expose Prometheus metrics."`
+}
+
+func init() {
+	metric.RegisterEmitter(&PrometheusConfig{})
+}
+
+func (config *PrometheusConfig) Description() string { return "Prometheus" }
+func (config *PrometheusConfig) IsConfigured() bool  { return config.Address != "" }
+
+func (config *PrometheusConfig) NewEmitter() (metric.Emitter, error) {
+	buildsStarted := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "started_total",
+		Help:      "Total number of Concourse builds started.",
+	})
+	prometheus.MustRegister(buildsStarted)
+
+	buildsFinished := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "finished_total",
+		Help:      "Total number of Concourse builds finished.",
+	})
+	prometheus.MustRegister(buildsFinished)
+
+	buildsSucceeded := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "succeeded_total",
+		Help:      "Total number of Concourse builds succeeded.",
+	})
+	prometheus.MustRegister(buildsSucceeded)
+
+	buildsErrored := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "errored_total",
+		Help:      "Total number of Concourse builds errored.",
+	})
+	prometheus.MustRegister(buildsErrored)
+
+	buildsFailed := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "failed_total",
+		Help:      "Total number of Concourse builds failed.",
+	})
+	prometheus.MustRegister(buildsFailed)
+
+	buildsAborted := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "aborted_total",
+		Help:      "Total number of Concourse builds aborted.",
+	})
+	prometheus.MustRegister(buildsAborted)
+
+	buildsFinishedVec := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "concourse",
+			Subsystem: "builds",
+			Name:      "finished",
+			Help:      "Count of builds finished across various dimensions.",
+		},
+		[]string{"team", "pipeline", "status"},
+	)
+	prometheus.MustRegister(buildsFinishedVec)
+	buildDurationsVec := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "concourse",
+			Subsystem: "builds",
+			Name:      "duration_seconds",
+			Help:      "Build time in seconds",
+		},
+		[]string{"team", "pipeline"},
+	)
+	prometheus.MustRegister(buildDurationsVec)
+
+	go func() {
+		http.Handle("/metrics", promhttp.Handler())
```

Comment:
Let's do something like this:

```go
listener, err := net.Listen("tcp", ...)
if err != nil {
  return nil, err
}

go http.Serve(listener, promhttp.Handler())

return &PromethiusEmitter{
// ...
```

This way the user will get an error if e.g. the port is in use, and the metrics endpoint won't be be slapped on to the global handler.

NB: this removes the `/metrics` routing, but I think it'll actually continue to work anyway; the handler itself probably isn't even looking at the path.

Reply:
Great idea!